### PR TITLE
feat: nextjs support

### PR DESCRIPTION
**Pull request by aidenybai**
**Please describe the changes this PR makes and why it should be merged:**

Next.js support!

TODO: docs
 
**Status**

- [x] Code changes have been tested against prettier, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the codebase
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
  - [x] This PR changes the internal workings with no modifications to the external API (bug fixes, performance improvements)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: packages/react/utils.ts [hunk 1]
```diff
@@ -22,14 +24,26 @@ export const unwrap = (vnode?: ReactNode): VNode => {
...
 export const flatten = <T>(rawChildren: T): T[] => {
   if (rawChildren === undefined || rawChildren === null) return [];
+  if (
+    typeof rawChildren === 'object' &&
+    'type' in rawChildren &&
+    'props' in rawChildren &&
+    rawChildren.props &&
+    typeof rawChildren.props === 'object' &&
+    'children' in rawChildren.props &&
+    rawChildren.type === FRAGMENT_SYMBOL
```
Comment (QuiiBz) [inline packages/react/utils.ts:43]: Do we need this many checks? We could only check the `type` property (and if `rawChildren` is an object) imo

Comment (tobySolutions) [reply packages/react/utils.ts:43]: Any updates on this?

Comment (aidenybai) [reply packages/react/utils.ts]: fixed, removed type generic

File: packages/compiler/react.ts [hunk 2]
```diff
@@ -133,13 +147,88 @@ const getDynamicsFromJSX = (
...
+      const createElement = addNamed(path, 'createElement', 'react', {
+        nameHint: 'createElement$',
+      });
+
+      const objectProperties: t.ObjectProperty[] = [];
+      for (const attribute of openingElement.attributes) {
+        if (t.isJSXAttribute(attribute)) {
+          if (
+            t.isJSXAttribute(attribute) &&
+            t.isJSXExpressionContainer(attribute.value)
```
Comment (SukkaW) [inline packages/compiler/react.ts:177]: Here `t.isJSXAttribute` is called twice.

File: website/pages/docs/quickstart.mdx [hunk 3]
```diff
@@ -72,7 +72,7 @@ Then, add the compiler to your bundler of choice:
   <Tab>
   ```js filename="rspack.config.js"
   // @ts-check
-  const million = require('million/compiler');
+  const million = require('million/compiler').default;
```
Comment (SukkaW) [inline website/pages/docs/quickstart.mdx:75]:
```suggestion
  const { default: million } = require('million/compiler');
```

This is my personal preference though and it is not blocking.

File: packages/next/src/next.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+// Must be in an input folder for mkdist to build
+import { createElement, useEffect, useReducer } from 'react';
+import type { ComponentProps, FunctionComponent } from 'react';
+
+let millionModule: any = null;
```
Comment (SukkaW) [inline packages/next/src/next.ts:5]:
```suggestion
let millionModule: typeof import('million/react') | null = null;
```

Does this type help?

Comment (aidenybai) [reply packages/next/src/next.ts:5]: wow i had no idea this was possible, thank you

File: packages/next/src/next.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      const importSource = async () => {
+        // @ts-expect-error - is defined
+        millionModule = await import('million/react');
+        blockFactory = millionModule.block(Component);
+      };
+      try {
+        void importSource();
+      } catch (e) {
+        throw new Error('Failed to load Million library');
+      }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/next/src/next.ts]: fixed

File: packages/compiler/react.ts (hunk 2 above)
Comment (aidenybai) [reply packages/compiler/react.ts]: Fixed

**Comment by aidenybai** [comment]
afaik ive tested most of nextjs compat. good idea to release as beta?

**Comment by QuiiBz** [comment]
> afaik ive tested most of nextjs compat. good idea to release as beta?

Everything seems fine to me!

**Review by QuiiBz — APPROVED** [review]

File: package.json [hunk 6]
```diff
@@ -74,7 +79,6 @@
   "dependencies": {
     "@babel/core": "^7.21.0",
     "@babel/plugin-syntax-jsx": "^7.21.4",
-    "@babel/plugin-syntax-typescript": "^7.21.4",
     "@babel/types": "^7.21.3",
     "forgetti": "^0.4.2",
```
Comment (SukkaW) [inline package.json:83]: Might wanna also bump `forgetti` here before releasing Next.js plugin. `forgetti` just fixed a bug a few hours ago: https://github.com/lxsmnsyc/forgetti/pull/4

Comment (aidenybai) [reply package.json:83]: good to know

**Pull request merged**